Modified `TranscriptProcessor` to use `TTSTextFrame`s

markbackman opened the pull request:
#### Please describe the changes in your PR. If it is addressing an issue, please reference that as well.

The motivation for this change is that on sessions ending, if the bot was interrupted, the bot's output would not be captured. `TTSTextFrame`s provide more frequent updates reflecting what the bot said—either word-level or sentence-level. This ensures that the transcript captures more of the bot's output.

This results in more accurate transcripts that better reflect what was actually spoken, especially in cases of interruption or early session termination.

This is a breaking change as it requires different placement of the processor:
```
pipeline = Pipeline(
            [
                transport.input(),
                stt,
                transcript.user(),              # User transcripts
                context_aggregator.user(),
                llm,
                tts,
                transport.output(),
                transcript.assistant_tts(),     # Assistant transcripts
                context_aggregator.assistant(),
            ]
        )
	
```

Maybe this is a case for an observer?

markbackman commented on src/pipecat/processors/transcript_processor.py:
```diff
@@ -64,89 +69,81 @@ async def process_frame(self, frame: Frame, direction: FrameDirection):
+            if content:
+                # Format timestamp with 3 decimal places
+                formatted_timestamp = (
+                    self._aggregation_start_time.strftime("%Y-%m-%dT%H:%M:%S.%f")[:-3] + "+00:00"
```
Generate a timestamp when aggregating text. This is close enough to reflect the actual timing.

markbackman commented on src/pipecat/processors/transcript_processor.py:
```diff
@@ -64,89 +69,81 @@ async def process_frame(self, frame: Frame, direction: FrameDirection):
+        if isinstance(frame, TTSTextFrame):
+            # Start timestamp on first text part
+            if not self._aggregation_start_time:
+                self._aggregation_start_time = datetime.now(timezone.utc)
```
Generate a timestamp when aggregating text. This is close enough to reflect the actual timing.

aconchillo replied:
We can use 

```
self._aggregation_start_time = time_now_iso8601()
```

markbackman replied:
I always forget about this utility...

markbackman commented on examples/foundational/28a-transcription-processor-openai.py:
```diff
@@ -130,6 +170,9 @@ async def on_transcript_update(processor, frame):
         @transport.event_handler("on_participant_left")
         async def on_participant_left(transport, participant, reason):
+            # Interrupt the TTS to stop the TTS generation
+            await task.queue_frame(StartInterruptionFrame())
```
I have to push a `StartInterruptionFrame` to get the TTS service to stop outputting audio and TTSTextFrames. I think this is likely an undesired behavior that we want to actually change.

aconchillo replied:
Oh, I missed this. yes, that seems not very convenient. Need to sleep on it, let's discuss tomorrow.

aconchillo commented on src/pipecat/processors/transcript_processor.py:
```diff
@@ -64,89 +69,81 @@ async def process_frame(self, frame: Frame, direction: FrameDirection):
+
+            self._current_text_parts.append(frame.text)
+
+        elif isinstance(frame, BotStoppedSpeakingFrame):
```
Maybe:

```
        elif isinstance(frame, (BotStoppedSpeakingFrame, StartInterruptionFrame)):
        
```

aconchillo commented on src/pipecat/processors/transcript_processor.py:
```diff
@@ -64,89 +69,81 @@ async def process_frame(self, frame: Frame, direction: FrameDirection):
+                # Format timestamp with 3 decimal places
+                formatted_timestamp = (
+                    self._aggregation_start_time.strftime("%Y-%m-%dT%H:%M:%S.%f")[:-3] + "+00:00"
+                )
```
already formatted with `time_now_iso8601()`

aconchillo commented on src/pipecat/processors/transcript_processor.py:
```diff
@@ -64,89 +69,81 @@ async def process_frame(self, frame: Frame, direction: FrameDirection):
+                message = TranscriptionMessage(
+                    role="assistant",
+                    content=content,
+                    timestamp=formatted_timestamp,
```
so, just use `self._aggregation_start_time:` here

aconchillo commented on src/pipecat/processors/transcript_processor.py:
```diff
@@ -64,89 +69,81 @@ async def process_frame(self, frame: Frame, direction: FrameDirection):
+        """Initialize processor with aggregation state."""
+        super().__init__(**kwargs)
+        self._current_text_parts: List[str] = []
+        self._aggregation_start_time: datetime | None = None
```
this will just be `Optional[str] = None`

aconchillo commented:
Minor comments. LGTM!

markbackman commented:
@aconchillo ok, changes made. Before I merge this, what do you think about:
> I have to push a StartInterruptionFrame to get the TTS service to stop outputting audio and TTSTextFrames. I think this is likely an undesired behavior that we want to actually change.

markbackman commented:
I've updated the `AssistantTranscriptProcessor` to emit the transcript callback when receiving a `CancelFrame`. I've also updated the examples to push a `CancelFrame` when the participant leaves.
